// @waves/ts-lib-crypto/dist/crypto/address-keys-seed.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.privateKey = exports.publicKey = exports.address = exports.keyPair = exports.buildAddress = exports.seedWithNonce = void 0;
const interface_1 = require("./interface");
const seed_1 = require("../extensions/seed");
const hashing_1 = require("./hashing");
const param_1 = require("../conversions/param");
const concat_split_1 = require("./concat-split");
const util_1 = require("./util");
const axlsign_1 = __importDefault(require("../libs/axlsign"));
const seedWithNonce = (seed, nonce) => ({ seed: seed_1.Seed.toBinary(seed).seed, nonce });
exports.seedWithNonce = seedWithNonce;
const buildAddress = (publicKeyBytes, chainId = interface_1.MAIN_NET_CHAIN_ID) => {
    const prefix = [1, typeof chainId === 'string' ? chainId.charCodeAt(0) : chainId];
    const publicKeyHashPart = (0, hashing_1._hashChain)(publicKeyBytes).slice(0, 20);
    const rawAddress = (0, concat_split_1.concat)(prefix, publicKeyHashPart);
    const addressHash = (0, hashing_1._hashChain)(rawAddress).slice(0, 4);
    return (0, concat_split_1.concat)(rawAddress, addressHash);
};
exports.buildAddress = buildAddress;
const buildSeedHash = (seedBytes, nonce) => {
    const nonceArray = [0, 0, 0, 0];
    if (nonce && nonce > 0) {
        let remainder = nonce;
        for (let i = 3; i >= 0; i--) {
            nonceArray[3 - i] = Math.floor(remainder / 2 ** (i * 8));
            remainder = remainder % 2 ** (i * 8);
        }
    }
    const seedBytesWithNonce = (0, concat_split_1.concat)(nonceArray, seedBytes);
    const seedHash = (0, hashing_1._hashChain)(seedBytesWithNonce);
    return (0, hashing_1.sha256)(seedHash);
};
const keyPair = (seed) => {
    const { seed: seedBytes, nonce } = seed_1.Seed.toBinary(seed);
    const seedHash = buildSeedHash(seedBytes, nonce);
    const keys = axlsign_1.default.generateKeyPair(seedHash);
    return {
        privateKey: keys.private,
        publicKey: keys.public,
    };
};
exports.keyPair = keyPair;
const address = (seedOrPublicKey, chainId = interface_1.MAIN_NET_CHAIN_ID) => (0, util_1.isPublicKey)(seedOrPublicKey) ?
    (0, exports.buildAddress)((0, param_1._fromIn)(seedOrPublicKey.publicKey), chainId) :
    (0, exports.address)((0, exports.keyPair)(seedOrPublicKey), chainId);
exports.address = address;
const publicKey = (seedOrPrivateKey) => (0, util_1.isPrivateKey)(seedOrPrivateKey) ?
    axlsign_1.default.generateKeyPair((0, param_1._fromIn)(seedOrPrivateKey.privateKey)).public :
    (0, exports.keyPair)(seedOrPrivateKey).publicKey;
exports.publicKey = publicKey;
const privateKey = (seed) => (0, exports.keyPair)(seed).privateKey;
exports.privateKey = privateKey;
