// @waves/ts-lib-crypto/dist/crypto/bls.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.blsVerify = exports.blsSign = exports.blsKeyPair = exports.blsPublicKey = void 0;
exports.mkBlsSecretKey = mkBlsSecretKey;
const bls12_381_js_1 = require("@noble/curves/bls12-381.js");
const hkdf_js_1 = require("@noble/hashes/hkdf.js");
const sha2_js_1 = require("@noble/hashes/sha2.js");
const utils_js_1 = require("@noble/curves/utils.js");
const param_1 = require("../conversions/param");
const DST = new TextEncoder().encode('BLS_SIG_BLS12381G2_XMD:SHA-256_SSWU_RO_NUL_');
const BLS_KEYGEN_SALT_BASE = new TextEncoder().encode('BLS-SIG-KEYGEN-SALT-');
const CURVE_ORDER = BigInt('0x73eda753299d7d483339d80809a1d80553bda402fffe5bfeffffffff00000001');
const bls = bls12_381_js_1.bls12_381.longSignatures;
function mkBlsSecretKey(seed) {
    const ikm = new Uint8Array([...seed, 0]);
    const info = new Uint8Array([0x00, 0x30]); // 48
    const okm = (0, hkdf_js_1.hkdf)(sha2_js_1.sha256, ikm, BLS_KEYGEN_SALT_BASE, info, 48);
    const sk = (0, utils_js_1.bytesToNumberBE)(okm) % CURVE_ORDER;
    if (sk === 0n)
        throw new Error('Invalid zero private key');
    const out = new Uint8Array(32);
    let tmp = sk;
    for (let i = 31; i >= 0; i--) {
        out[i] = Number(tmp & 0xffn);
        tmp >>= 8n;
    }
    return out;
}
function hashToPoint(message) {
    return bls12_381_js_1.bls12_381.G2.hashToCurve(message, { DST });
}
const blsPublicKey = (secretKey) => {
    return bls.getPublicKey((0, param_1._fromIn)(secretKey)).toBytes(true);
};
exports.blsPublicKey = blsPublicKey;
const blsKeyPair = (seed) => {
    const secretKey = mkBlsSecretKey((0, param_1._fromIn)(seed));
    const publicKey = (0, exports.blsPublicKey)(secretKey);
    return { blsSecret: secretKey, blsPublic: publicKey };
};
exports.blsKeyPair = blsKeyPair;
const blsSign = (secretKey, message) => {
    const msgPoint = hashToPoint((0, param_1._fromIn)(message));
    return bls.sign(msgPoint, (0, param_1._fromIn)(secretKey)).toBytes(true);
};
exports.blsSign = blsSign;
const blsVerify = (publicKey, message, signature) => {
    const msgPoint = hashToPoint((0, param_1._fromIn)(message));
    return bls.verify((0, param_1._fromIn)(signature), msgPoint, (0, param_1._fromIn)(publicKey));
};
exports.blsVerify = blsVerify;
