// @waves/ts-lib-crypto/dist/crypto/crypto.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.crypto = void 0;
const random_1 = require("./random");
const encryption_1 = require("./encryption");
const base_xx_1 = require("../conversions/base-xx");
const string_bytes_1 = require("../conversions/string-bytes");
const concat_split_1 = require("./concat-split");
const hashing_1 = require("./hashing");
const address_keys_seed_1 = require("./address-keys-seed");
const sign_1 = require("./sign");
const verification_1 = require("./verification");
const rsa_1 = require("./rsa");
const seed_ecryption_1 = require("./seed-ecryption");
const merkle_verify_1 = require("./merkle-verify");
const bls_1 = require("./bls");
const crypto = (options) => {
    if (options && options.seed == '')
        throw new Error('Empty seed is not allowed.');
    const c = (f, first) => (...args) => f(first, ...args);
    const toOut = (f) => (...args) => {
        const r = f(...args);
        if (!options || options.output === 'Base58') {
            return (0, base_xx_1.base58Encode)(r);
        }
        else {
            return r;
        }
    };
    const toOutKey = (f) => (...args) => {
        const r = f(...args);
        return ((!options || options && options.output === 'Base58') ?
            ({ privateKey: (0, base_xx_1.base58Encode)(r.privateKey), publicKey: (0, base_xx_1.base58Encode)(r.publicKey) }) :
            r);
    };
    const s = (options && options.seed) ? options.seed : undefined;
    const seedPart = {
        seedWithNonce: s ? c(address_keys_seed_1.seedWithNonce, s) : address_keys_seed_1.seedWithNonce,
        signBytes: toOut(s ? c(sign_1.signBytes, s) : sign_1.signBytes),
        keyPair: toOutKey(s ? c(address_keys_seed_1.keyPair, s) : address_keys_seed_1.keyPair),
        publicKey: toOut(s ? c(address_keys_seed_1.publicKey, s) : address_keys_seed_1.publicKey),
        privateKey: toOut(s ? c(address_keys_seed_1.privateKey, s) : address_keys_seed_1.privateKey),
        address: toOut(s ? c(address_keys_seed_1.address, s) : address_keys_seed_1.address),
    };
    return {
        ...seedPart,
        sharedKey: toOut(encryption_1.sharedKey),
        buildAddress: address_keys_seed_1.buildAddress,
        blake2b: hashing_1.blake2b,
        keccak: hashing_1.keccak,
        sha256: hashing_1.sha256,
        base64Encode: base_xx_1.base64Encode,
        base64Decode: base_xx_1.base64Decode,
        base58Encode: base_xx_1.base58Encode,
        base58Decode: base_xx_1.base58Decode,
        base16Encode: base_xx_1.base16Encode,
        base16Decode: base_xx_1.base16Decode,
        stringToBytes: string_bytes_1.stringToBytes,
        bytesToString: string_bytes_1.bytesToString,
        random: random_1.random,
        randomSeed: random_1.randomSeed,
        randomBytes: random_1.randomBytes,
        verifySignature: verification_1.verifySignature,
        verifyPublicKey: verification_1.verifyPublicKey,
        verifyAddress: verification_1.verifyAddress,
        messageDecrypt: encryption_1.messageDecrypt,
        messageEncrypt: encryption_1.messageEncrypt,
        aesDecrypt: encryption_1.aesDecrypt,
        aesEncrypt: encryption_1.aesEncrypt,
        encryptSeed: seed_ecryption_1.encryptSeed,
        decryptSeed: seed_ecryption_1.decryptSeed,
        rsaKeyPair: rsa_1.rsaKeyPair,
        rsaKeyPairSync: rsa_1.rsaKeyPairSync,
        rsaSign: rsa_1.rsaSign,
        rsaVerify: rsa_1.rsaVerify,
        merkleVerify: merkle_verify_1.merkleVerify,
        split: concat_split_1.split,
        concat: concat_split_1.concat,
        blsKeyPair: bls_1.blsKeyPair,
        blsPublicKey: bls_1.blsPublicKey,
        blsSign: bls_1.blsSign,
        blsVerify: bls_1.blsVerify,
    };
};
exports.crypto = crypto;
