// @waves/marshall/dist/parsePrimitives.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.byteToScript = exports.byteNewAliasToString = exports.byteToAddressOrAlias = exports.byteToBase58WithLength = exports.byteToBase58 = exports.byteToStringWithLength = exports.P_BASE64 = exports.P_BASE58_VAR = exports.P_BASE58_FIXED = exports.P_STRING_VAR = exports.P_STRING_FIXED = exports.P_BOOLEAN = exports.P_LONG = exports.P_INT = exports.P_SHORT = exports.P_BYTE = exports.P_OPTION = exports.ALIAS_VERSION = void 0;
const long_1 = __importDefault(require("long"));
const Utf8ArrayToStr_1 = require("./libs/Utf8ArrayToStr");
const base58_1 = __importDefault(require("./libs/base58"));
const Base64 = __importStar(require("base64-js"));
exports.ALIAS_VERSION = 2;
const LENGTH_SIZE = 2;
const P_OPTION = (p) => (bytes, start = 0) => {
    if (bytes[start] === 0)
        return { value: null, shift: 1 };
    const result = p(bytes, start + 1);
    return { value: result.value, shift: result.shift + 1 };
};
exports.P_OPTION = P_OPTION;
const P_BYTE = (bytes, start = 0) => ({ value: bytes[start], shift: 1 });
exports.P_BYTE = P_BYTE;
const P_SHORT = (bytes, start = 0) => ({ value: 256 * bytes[start] + bytes[start + 1], shift: 2 });
exports.P_SHORT = P_SHORT;
const P_INT = (bytes, start = 0) => ({
    value: Math.pow(2, 24) * bytes[start] + Math.pow(2, 16) * bytes[start + 1] + Math.pow(2, 8) * bytes[start + 2] + bytes[start + 3],
    shift: 4
});
exports.P_INT = P_INT;
const P_LONG = (bytes, start = 0) => ({
    value: long_1.default.fromBytesBE(Array.from(bytes.slice(start, start + 8))).toString(),
    shift: 8,
});
exports.P_LONG = P_LONG;
const P_BOOLEAN = (bytes, start = 0) => {
    const value = !!bytes[start];
    return { value, shift: 1 };
};
exports.P_BOOLEAN = P_BOOLEAN;
const P_STRING_FIXED = (len) => (bytes, start = 0) => {
    const value = (0, Utf8ArrayToStr_1.Utf8ArrayToStr)(bytes.slice(start, start + len));
    return { shift: len, value };
};
exports.P_STRING_FIXED = P_STRING_FIXED;
const P_STRING_VAR = (lenParser) => (bytes, start = 0) => {
    const lengthInfo = lenParser(bytes, start);
    const { value } = (0, exports.P_STRING_FIXED)(lengthInfo.value)(bytes, start + lengthInfo.shift);
    return { shift: lengthInfo.value + lengthInfo.shift, value };
};
exports.P_STRING_VAR = P_STRING_VAR;
const P_BASE58_FIXED = (len) => (bytes, start = 0) => {
    const value = base58_1.default.encode(bytes.slice(start, start + len));
    return { value, shift: len };
};
exports.P_BASE58_FIXED = P_BASE58_FIXED;
const P_BASE58_VAR = (lenParser) => (bytes, start = 0) => {
    const lengthInfo = lenParser(bytes, start);
    const { value } = (0, exports.P_BASE58_FIXED)(lengthInfo.value)(bytes, start + LENGTH_SIZE);
    return { shift: lengthInfo.value + LENGTH_SIZE, value };
};
exports.P_BASE58_VAR = P_BASE58_VAR;
const P_BASE64 = (lenParser) => (bytes, start = 0) => {
    const lengthInfo = lenParser(bytes, start);
    const value = `base64:${Base64.fromByteArray(bytes.slice(start + lengthInfo.shift, start + lengthInfo.shift + lengthInfo.value))}`;
    return { shift: lengthInfo.value + lengthInfo.shift, value };
};
exports.P_BASE64 = P_BASE64;
const byteToString = (shift) => (bytes, start) => {
    const value = (0, Utf8ArrayToStr_1.Utf8ArrayToStr)(bytes.slice(start, start + shift));
    return { shift, value };
};
const byteToStringWithLength = (bytes, start = 0) => {
    const lengthInfo = (0, exports.P_SHORT)(bytes, start);
    const { value } = byteToString(lengthInfo.value)(bytes, start + LENGTH_SIZE);
    return { shift: lengthInfo.value + LENGTH_SIZE, value };
};
exports.byteToStringWithLength = byteToStringWithLength;
const byteToBase58 = (bytes, start = 0, length) => {
    const shift = length || 32;
    const value = base58_1.default.encode(bytes.slice(start, start + shift));
    return { value, shift };
};
exports.byteToBase58 = byteToBase58;
const byteToBase58WithLength = (bytes, start = 0) => {
    const lenInfo = (0, exports.P_SHORT)(bytes, start);
    const value = base58_1.default.encode(bytes.slice(start + lenInfo.shift, start + lenInfo.shift + lenInfo.value));
    return { value, shift: lenInfo.shift + lenInfo.value };
};
exports.byteToBase58WithLength = byteToBase58WithLength;
const byteToAddressOrAlias = (bytes, start = 0) => {
    if (bytes[start] === exports.ALIAS_VERSION) {
        const aliasData = (0, exports.byteToStringWithLength)(bytes, start + 2);
        return { shift: aliasData.shift + 2, value: `alias:${String.fromCharCode(bytes[start + 1])}:${aliasData.value}` };
    }
    else {
        return (0, exports.byteToBase58)(bytes, start, 26);
    }
};
exports.byteToAddressOrAlias = byteToAddressOrAlias;
const byteNewAliasToString = (bytes, start = 0) => {
    const shift = (0, exports.P_SHORT)(bytes, start).value + LENGTH_SIZE;
    const { value } = (0, exports.byteToStringWithLength)(bytes, start);
    return { shift, value };
};
exports.byteNewAliasToString = byteNewAliasToString;
const byteToScript = (bytes, start = 0) => {
    const VERSION_LENGTH = 1;
    if (bytes[start] === 0) {
        return { shift: VERSION_LENGTH, value: null };
    }
    const lengthInfo = (0, exports.P_SHORT)(bytes, start + VERSION_LENGTH);
    const from = start + VERSION_LENGTH + lengthInfo.shift;
    const to = start + VERSION_LENGTH + lengthInfo.shift + lengthInfo.value;
    const value = `base64:${Base64.fromByteArray(bytes.slice(from, to))}`;
    return { value, shift: to - start };
};
exports.byteToScript = byteToScript;
