// @waves/marshall/dist/serializePrimitives.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ADDRESS_OR_ALIAS = exports.ALIAS = exports.SCRIPT = exports.LONG = exports.COUNT = exports.LEN = exports.OPTION = exports.INT = exports.SHORT = exports.BYTES = exports.BOOL = exports.BYTE = exports.STRING = exports.BASE64_STRING = exports.BASE58_STRING = exports.one = exports.zero = exports.empty = void 0;
const base58_1 = __importDefault(require("./libs/base58"));
const Base64 = __importStar(require("base64-js"));
const utils_1 = require("./libs/utils");
const long_1 = __importDefault(require("long"));
const stringToUint8Array = (str) => Uint8Array.from([...unescape(encodeURIComponent(str))].map(c => c.charCodeAt(0)));
exports.empty = Uint8Array.from([]);
exports.zero = Uint8Array.from([0]);
exports.one = Uint8Array.from([1]);
const BASE58_STRING = (value) => base58_1.default.decode(value);
exports.BASE58_STRING = BASE58_STRING;
const BASE64_STRING = (value) => Base64.toByteArray(value.replace('base64:', ''));
exports.BASE64_STRING = BASE64_STRING;
const STRING = (value) => value ? stringToUint8Array(value) : exports.empty;
exports.STRING = STRING;
const BYTE = (value) => Uint8Array.from([value]);
exports.BYTE = BYTE;
const BOOL = (value) => (0, exports.BYTE)(value == true ? 1 : 0);
exports.BOOL = BOOL;
const BYTES = (value) => Uint8Array.from(value);
exports.BYTES = BYTES;
const SHORT = (value) => {
    const s = long_1.default.fromNumber(value, true);
    return Uint8Array.from(s.toBytesBE().slice(6));
};
exports.SHORT = SHORT;
const INT = (value) => {
    const i = long_1.default.fromNumber(value, true);
    return Uint8Array.from(i.toBytesBE().slice(4));
};
exports.INT = INT;
const OPTION = (s) => (value) => value == null
    || (typeof value == 'string' && value.length == 0)
    ? exports.zero : (0, utils_1.concat)(exports.one, s(value));
exports.OPTION = OPTION;
const LEN = (lenSerializer) => (valueSerializer) => (value) => {
    const data = valueSerializer(value);
    const len = lenSerializer(data.length);
    return (0, utils_1.concat)(len, data);
};
exports.LEN = LEN;
const COUNT = (countSerializer) => (itemSerializer) => (items) => {
    const data = (0, utils_1.concat)(...items.map(x => itemSerializer(x)));
    const len = countSerializer(items.length);
    return (0, utils_1.concat)(len, data);
};
exports.COUNT = COUNT;
const LONG = (value) => {
    let l;
    if (typeof value === 'number') {
        if (value > Math.pow(2, 53) - 1) {
            throw new Error(`${value} is too big to be precisely represented as js number. Use string instead`);
        }
        l = long_1.default.fromNumber(value);
    }
    else {
        l = long_1.default.fromString(value.toString());
    }
    return Uint8Array.from(l.toBytesBE());
};
exports.LONG = LONG;
const SCRIPT = (script) => (0, exports.OPTION)((0, exports.LEN)(exports.SHORT)(exports.BASE64_STRING))(script ? script.slice(7) : null);
exports.SCRIPT = SCRIPT;
const ALIAS = val => {
    const [_, byte, alias] = val.split(':');
    if (!byte || byte.length !== 1)
        throw new Error('Invalid network byte in alias');
    if (!alias || alias.length === 0)
        throw new Error('Invalid alias body');
    return (0, utils_1.concat)([2], [byte.charCodeAt(0)], (0, exports.LEN)(exports.SHORT)(exports.STRING)(alias));
};
exports.ALIAS = ALIAS;
const ADDRESS_OR_ALIAS = val => val.startsWith('alias')
    ? (0, exports.ALIAS)(val)
    : (0, exports.BASE58_STRING)(val);
exports.ADDRESS_OR_ALIAS = ADDRESS_OR_ALIAS;
