// @waves/ts-lib-crypto/dist/conversions/base-xx.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.base16Encode = exports.base16Decode = exports.base58Encode = exports.base58Decode = exports.base64Encode = exports.base64Decode = void 0;
//@ts-ignore
const util_1 = require("node-forge/lib/util");
const base58_1 = __importDefault(require("../libs/base58"));
const param_1 = require("./param");
const string_bytes_1 = require("./string-bytes");
const base64Decode = (input) => (0, string_bytes_1.stringToBytes)((0, util_1.decode64)(input), 'raw');
exports.base64Decode = base64Decode;
const base64Encode = (input) => (0, util_1.encode64)((0, string_bytes_1.bytesToString)((0, param_1._fromIn)(input), 'raw'));
exports.base64Encode = base64Encode;
const base58Decode = (input) => base58_1.default.decode(input);
exports.base58Decode = base58Decode;
const base58Encode = (input) => base58_1.default.encode((0, param_1._fromIn)(input));
exports.base58Encode = base58Encode;
const base16Decode = (input) => (0, string_bytes_1.stringToBytes)((0, util_1.hexToBytes)(input), 'raw');
exports.base16Decode = base16Decode;
const base16Encode = (input) => (0, util_1.bytesToHex)((0, string_bytes_1.bytesToString)((0, param_1._fromIn)(input), 'raw'));
exports.base16Encode = base16Encode;
