// @waves/waves-transactions/dist/general.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.txTypeMap = void 0;
exports.signTx = signTx;
exports.serialize = serialize;
exports.verify = verify;
exports.verifyCustomData = verifyCustomData;
exports.verifyAuthData = verifyAuthData;
exports.verifyWavesAuthData = verifyWavesAuthData;
const marshall_1 = require("@waves/marshall");
const ts_lib_crypto_1 = require("@waves/ts-lib-crypto");
const ts_types_1 = require("@waves/ts-types");
const proto_serialize_1 = require("./proto-serialize");
const defaultVersions_1 = require("./defaultVersions");
const issue_1 = require("./transactions/issue");
const transfer_1 = require("./transactions/transfer");
const reissue_1 = require("./transactions/reissue");
const burn_1 = require("./transactions/burn");
const lease_1 = require("./transactions/lease");
const cancel_lease_1 = require("./transactions/cancel-lease");
const data_1 = require("./transactions/data");
const mass_transfer_1 = require("./transactions/mass-transfer");
const alias_1 = require("./transactions/alias");
const set_script_1 = require("./transactions/set-script");
const generic_1 = require("./generic");
const set_asset_script_1 = require("./transactions/set-asset-script");
const exchange_1 = require("./transactions/exchange");
const sponsorship_1 = require("./transactions/sponsorship");
const invoke_script_1 = require("./transactions/invoke-script");
const custom_data_1 = require("./requests/custom-data");
const auth_1 = require("./requests/auth");
const wavesAuth_1 = require("./requests/wavesAuth");
const commit_to_generation_1 = require("./transactions/commit-to-generation");
const update_asset_info_1 = require("./transactions/update-asset-info");
exports.txTypeMap = {
    [ts_types_1.TRANSACTION_TYPE.ISSUE]: { sign: (x, seed) => (0, issue_1.issue)(x, seed) },
    [ts_types_1.TRANSACTION_TYPE.TRANSFER]: { sign: (x, seed) => (0, transfer_1.transfer)(x, seed) },
    [ts_types_1.TRANSACTION_TYPE.REISSUE]: { sign: (x, seed) => (0, reissue_1.reissue)(x, seed) },
    [ts_types_1.TRANSACTION_TYPE.BURN]: { sign: (x, seed) => (0, burn_1.burn)(x, seed) },
    [ts_types_1.TRANSACTION_TYPE.LEASE]: { sign: (x, seed) => (0, lease_1.lease)(x, seed) },
    [ts_types_1.TRANSACTION_TYPE.CANCEL_LEASE]: { sign: (x, seed) => (0, cancel_lease_1.cancelLease)(x, seed) },
    [ts_types_1.TRANSACTION_TYPE.ALIAS]: { sign: (x, seed) => (0, alias_1.alias)(x, seed) },
    [ts_types_1.TRANSACTION_TYPE.MASS_TRANSFER]: { sign: (x, seed) => (0, mass_transfer_1.massTransfer)(x, seed) },
    [ts_types_1.TRANSACTION_TYPE.DATA]: { sign: (x, seed) => (0, data_1.data)(x, seed) },
    [ts_types_1.TRANSACTION_TYPE.SET_SCRIPT]: { sign: (x, seed) => (0, set_script_1.setScript)(x, seed) },
    [ts_types_1.TRANSACTION_TYPE.SET_ASSET_SCRIPT]: { sign: (x, seed) => (0, set_asset_script_1.setAssetScript)(x, seed) },
    [ts_types_1.TRANSACTION_TYPE.SPONSORSHIP]: { sign: (x, seed) => (0, sponsorship_1.sponsorship)(x, seed) },
    [ts_types_1.TRANSACTION_TYPE.EXCHANGE]: { sign: (x, seed) => (0, exchange_1.exchange)(x, seed) },
    [ts_types_1.TRANSACTION_TYPE.INVOKE_SCRIPT]: { sign: (x, seed) => (0, invoke_script_1.invokeScript)(x, seed) },
    [ts_types_1.TRANSACTION_TYPE.UPDATE_ASSET_INFO]: { sign: (x, seed) => (0, update_asset_info_1.updateAssetInfo)(x, seed) },
    [ts_types_1.TRANSACTION_TYPE.COMMIT_TO_GENERATION]: { sign: (x, seed) => (0, commit_to_generation_1.commitToGeneration)(x, seed) },
    // [TRANSACTION_TYPE.INVOKE_EXPRESSION]: {sign: (x, seed) => invokeExpression(x as InvokeExpressionTransaction, seed)},
};
const latestVersionByType = {
    [ts_types_1.TRANSACTION_TYPE.ISSUE]: defaultVersions_1.DEFAULT_VERSIONS.ISSUE,
    [ts_types_1.TRANSACTION_TYPE.TRANSFER]: defaultVersions_1.DEFAULT_VERSIONS.TRANSFER,
    [ts_types_1.TRANSACTION_TYPE.REISSUE]: defaultVersions_1.DEFAULT_VERSIONS.REISSUE,
    [ts_types_1.TRANSACTION_TYPE.BURN]: defaultVersions_1.DEFAULT_VERSIONS.BURN,
    [ts_types_1.TRANSACTION_TYPE.EXCHANGE]: defaultVersions_1.DEFAULT_VERSIONS.EXCHANGE,
    [ts_types_1.TRANSACTION_TYPE.LEASE]: defaultVersions_1.DEFAULT_VERSIONS.LEASE,
    [ts_types_1.TRANSACTION_TYPE.CANCEL_LEASE]: defaultVersions_1.DEFAULT_VERSIONS.CANCEL_LEASE,
    [ts_types_1.TRANSACTION_TYPE.ALIAS]: defaultVersions_1.DEFAULT_VERSIONS.ALIAS,
    [ts_types_1.TRANSACTION_TYPE.MASS_TRANSFER]: defaultVersions_1.DEFAULT_VERSIONS.MASS_TRANSFER,
    [ts_types_1.TRANSACTION_TYPE.DATA]: defaultVersions_1.DEFAULT_VERSIONS.DATA,
    [ts_types_1.TRANSACTION_TYPE.SET_SCRIPT]: defaultVersions_1.DEFAULT_VERSIONS.SET_SCRIPT,
    [ts_types_1.TRANSACTION_TYPE.SPONSORSHIP]: defaultVersions_1.DEFAULT_VERSIONS.SPONSORSHIP,
    [ts_types_1.TRANSACTION_TYPE.SET_ASSET_SCRIPT]: defaultVersions_1.DEFAULT_VERSIONS.SET_ASSET_SCRIPT,
    [ts_types_1.TRANSACTION_TYPE.INVOKE_SCRIPT]: defaultVersions_1.DEFAULT_VERSIONS.INVOKE_SCRIPT,
    [ts_types_1.TRANSACTION_TYPE.UPDATE_ASSET_INFO]: defaultVersions_1.DEFAULT_VERSIONS.UPDATE_ASSET_INFO,
    [ts_types_1.TRANSACTION_TYPE.COMMIT_TO_GENERATION]: defaultVersions_1.DEFAULT_VERSIONS.COMMIT_TO_GENERATION,
};
/**
 * Signs arbitrary transaction. Can also create signed transaction if provided params have type field
 * @param tx
 * @param seed
 */
function signTx(tx, seed) {
    if (!exports.txTypeMap[tx.type])
        throw new Error(`Unknown tx type: ${tx.type}`);
    return exports.txTypeMap[tx.type].sign(tx, seed);
}
/**
 * Converts transaction or order object to Uint8Array
 * @param obj transaction or order
 */
function serialize(obj) {
    if ((0, generic_1.isOrder)(obj))
        return marshall_1.binary.serializeOrder(obj);
    const tx = obj;
    if (latestVersionByType[tx.type] === tx.version)
        return (0, proto_serialize_1.txToProtoBytes)(tx);
    return marshall_1.binary.serializeTx(obj);
}
/**
 * Verifies signature of transaction or order
 * @param obj
 * @param proofN - proof index. Takes first proof by default
 * @param publicKey - takes senderPublicKey by default
 */
function verify(obj, proofN = 0, publicKey) {
    publicKey = publicKey || obj.senderPublicKey;
    const bytes = serialize(obj);
    const signature = obj.version == null ? obj.signature : obj.proofs[proofN];
    return (0, ts_lib_crypto_1.verifySignature)(publicKey, bytes, signature);
}
function verifyCustomData(data) {
    const bytes = (0, custom_data_1.serializeCustomData)(data);
    return (0, ts_lib_crypto_1.verifySignature)(data.publicKey, bytes, data.signature);
}
function verifyAuthData(authData, params, chainId) {
    chainId = chainId || 'W';
    const bytes = (0, auth_1.serializeAuthData)(params);
    const myAddress = (0, ts_lib_crypto_1.address)({ publicKey: authData.publicKey }, chainId);
    return myAddress === authData.address && (0, ts_lib_crypto_1.verifySignature)(authData.publicKey, bytes, authData.signature);
}
function verifyWavesAuthData(authData, params, chainId) {
    chainId = chainId || 'W';
    const bytes = (0, wavesAuth_1.serializeWavesAuthData)(params);
    const myAddress = (0, ts_lib_crypto_1.address)({ publicKey: authData.publicKey }, chainId);
    return myAddress === authData.address && (0, ts_lib_crypto_1.verifySignature)(authData.publicKey, bytes, authData.signature);
}
