// @waves/waves-transactions/dist/proto-serialize.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.scriptToProto = exports.dataEntryToProto = exports.signedTxToProto = exports.txToProto = void 0;
exports.txToProtoBytes = txToProtoBytes;
exports.signedTxToProtoBytes = signedTxToProtoBytes;
exports.protoBytesToSignedTx = protoBytesToSignedTx;
exports.protoBytesToTx = protoBytesToTx;
exports.protoTxDataToTx = protoTxDataToTx;
exports.orderToProtoBytes = orderToProtoBytes;
exports.protoBytesToOrder = protoBytesToOrder;
const wavesProto = __importStar(require("@waves/protobuf-serialization"));
const ts_lib_crypto_1 = require("@waves/ts-lib-crypto");
const marshall_1 = require("@waves/marshall");
const ts_types_1 = require("@waves/ts-types");
const generic_1 = require("./generic");
const long_1 = __importDefault(require("long"));
const invokeScriptCallSchema = {
    ...marshall_1.schemas.txFields.functionCall[1],
};
const recipientFromProto = (recipient, chainId) => {
    if (recipient.alias) {
        return `alias:${String.fromCharCode(chainId)}:${recipient.alias}`;
    }
    const rawAddress = (0, ts_lib_crypto_1.concat)([1], [chainId], recipient.publicKeyHash);
    const checkSum = (0, ts_lib_crypto_1.keccak)((0, ts_lib_crypto_1.blake2b)(rawAddress)).slice(0, 4);
    return (0, ts_lib_crypto_1.base58Encode)((0, ts_lib_crypto_1.concat)(rawAddress, checkSum));
};
function convertNumber(n) {
    const maxJsNumber = 2 ** 53 - 1;
    return n.toNumber() > maxJsNumber ? n.toString() : n.toNumber();
}
function txToProtoBytes(obj) {
    return new Uint8Array(wavesProto.waves.Transaction.encode((0, exports.txToProto)(obj)).finish());
}
function signedTxToProtoBytes(obj) {
    return new Uint8Array(wavesProto.waves.SignedTransaction.encode((0, exports.signedTxToProto)(obj)).finish());
}
function protoBytesToSignedTx(bytes) {
    const txData = wavesProto.waves.SignedTransaction.decode(bytes);
    const tx = protoTxDataToTx(txData.transaction);
    return {
        ...tx,
        proofs: (txData.proofs || []).map(uint8Array2proof),
    };
}
function protoBytesToTx(bytes) {
    const t = wavesProto.waves.Transaction.decode(bytes);
    return protoTxDataToTx(t);
}
function protoTxDataToTx(t) {
    // | 'invokeExpression'
    let res = {
        version: t.version,
        type: typeByName[t.data],
        senderPublicKey: (0, ts_lib_crypto_1.base58Encode)(t.senderPublicKey),
        timestamp: t.timestamp.toNumber(),
        fee: convertNumber(t.fee.amount),
        // chainId: t.chainId
    };
    if (Object.hasOwn(t.fee, 'assetId')) {
        res.feeAssetId = (0, ts_lib_crypto_1.base58Encode)(t.fee.assetId);
    }
    else {
        res.feeAssetId = null;
    }
    if (Object.hasOwn(t, 'chainId')) {
        res.chainId = t.chainId;
    }
    switch (t.data) {
        case 'issue':
            res.name = t.issue.name;
            res.description = t.issue.description;
            res.quantity = convertNumber(t.issue.amount);
            res.decimals = t.issue.decimals;
            res.reissuable = t.issue.reissuable;
            res.script = (Object.hasOwn(t.issue, 'script')) ? (0, generic_1.base64Prefix)((0, ts_lib_crypto_1.base64Encode)(t.issue.script)) : null;
            break;
        case 'transfer':
            res.amount = convertNumber(t.transfer.amount.amount);
            res.recipient = recipientFromProto(t.transfer.recipient, t.chainId);
            res.attachment = (Object.hasOwn(t.transfer, 'attachment')) ? (0, ts_lib_crypto_1.base58Encode)(t.transfer.attachment) : '';
            res.assetId = (Object.hasOwn(t.transfer.amount, 'assetId')) ? (0, ts_lib_crypto_1.base58Encode)(t.transfer.amount.assetId) : null;
            break;
        case 'reissue':
            res.quantity = convertNumber(t.reissue.assetAmount.amount);
            res.assetId = t.reissue.assetAmount.assetId == null ? null : (0, ts_lib_crypto_1.base58Encode)(t.reissue.assetAmount.assetId);
            res.reissuable = t.reissue.reissuable;
            break;
        case 'burn':
            res.amount = convertNumber(t.burn.assetAmount.amount);
            res.assetId = (0, ts_lib_crypto_1.base58Encode)(t.burn.assetAmount.assetId);
            break;
        case 'exchange':
            res.amount = convertNumber(t.exchange.amount);
            res.price = convertNumber(t.exchange.price);
            res.buyMatcherFee = convertNumber(t.exchange.buyMatcherFee);
            res.sellMatcherFee = convertNumber(t.exchange.sellMatcherFee);
            res.order1 = orderFromProto(t.exchange.orders[0]);
            res.order2 = orderFromProto(t.exchange.orders[1]);
            break;
        case 'lease':
            res.recipient = recipientFromProto(t.lease.recipient, t.chainId);
            res.amount = convertNumber(t.lease.amount);
            break;
        case 'leaseCancel':
            res.leaseId = (0, ts_lib_crypto_1.base58Encode)(t.leaseCancel.leaseId);
            break;
        case 'createAlias':
            res.alias = t.createAlias.alias;
            break;
        case 'massTransfer':
            res.assetId = (Object.hasOwn(t.massTransfer, 'assetId')) ? (0, ts_lib_crypto_1.base58Encode)(t.massTransfer.assetId) : null;
            res.attachment = (Object.hasOwn(t.massTransfer, 'attachment')) ? (0, ts_lib_crypto_1.base58Encode)(t.massTransfer.attachment) : '';
            res.transfers = t.massTransfer.transfers.map(({ amount, recipient }) => ({
                amount: convertNumber(amount),
                recipient: recipientFromProto(recipient, t.chainId),
            }));
            break;
        case 'dataTransaction':
            res.data = t.dataTransaction.data.map(de => {
                if (Object.hasOwn(de, 'binaryValue'))
                    return {
                        key: de.key,
                        type: 'binary',
                        value: (0, generic_1.base64Prefix)((0, ts_lib_crypto_1.base64Encode)(de.binaryValue)),
                    };
                if (Object.hasOwn(de, 'boolValue'))
                    return { key: de.key, type: 'boolean', value: de.boolValue };
                if (Object.hasOwn(de, 'intValue'))
                    return {
                        key: de.key,
                        type: 'integer',
                        value: convertNumber(de.intValue),
                    };
                if (Object.hasOwn(de, 'stringValue'))
                    return { key: de.key, type: 'string', value: de.stringValue };
                return { key: de.key, value: null };
            });
            break;
        case 'setScript':
            res.script = (Object.hasOwn(t.setScript, 'script')) ? (0, generic_1.base64Prefix)((0, ts_lib_crypto_1.base64Encode)(t.setScript.script)) : null;
            break;
        case 'sponsorFee':
            res.minSponsoredAssetFee = convertNumber(t.sponsorFee.minFee.amount);
            res.assetId = (0, ts_lib_crypto_1.base58Encode)(t.sponsorFee.minFee.assetId);
            break;
        case 'setAssetScript':
            res.assetId = (0, ts_lib_crypto_1.base58Encode)(t.setAssetScript.assetId);
            res.script = (0, generic_1.base64Prefix)((0, ts_lib_crypto_1.base64Encode)(t.setAssetScript.script));
            break;
        case 'invokeScript':
            res.dApp = recipientFromProto(t.invokeScript.dApp, t.chainId);
            if (t.invokeScript.functionCall != null) {
                res.call = marshall_1.binary.parserFromSchema(invokeScriptCallSchema)(t.invokeScript.functionCall).value; //todo: export function call from marshall and use it directly
            }
            res.payment = t.invokeScript.payments.map(p => ({
                amount: convertNumber(p.amount),
                assetId: Object.hasOwn(p, 'assetId') ? (0, ts_lib_crypto_1.base58Encode)(p.assetId) : null,
            }));
            break;
        case 'updateAssetInfo':
            res.assetId = (0, ts_lib_crypto_1.base58Encode)(t.updateAssetInfo.assetId);
            res.name = t.updateAssetInfo.name;
            res.description = t.updateAssetInfo.description;
            break;
        case 'commitToGeneration':
            res.generationPeriodStart = t.commitToGeneration.generationPeriodStart;
            res.endorserPublicKey = t.commitToGeneration.endorserPublicKey == null || t.commitToGeneration.endorserPublicKey.length === 0
                ? null
                : (0, ts_lib_crypto_1.base58Encode)(t.commitToGeneration.endorserPublicKey);
            res.commitmentSignature = t.commitToGeneration.commitmentSignature == null || t.commitToGeneration.commitmentSignature.length === 0
                ? null
                : (0, ts_lib_crypto_1.base58Encode)(t.commitToGeneration.commitmentSignature);
            break;
        // case 'invokeExpression':
        //     res.expression = t.invokeExpression?.expression == null ? null : base64Prefix(base64Encode(t.invokeExpression?.expression))
        //     break
        default:
            throw new Error(`Unsupported tx type ${t.data}`);
    }
    if (Object.hasOwn(res, 'chainId')) {
        res.sender = (0, ts_lib_crypto_1.address)({ publicKey: t.senderPublicKey }, t.chainId);
    }
    else {
        let recipient = res.recipient || res.dApp || (res.transfers && res.transfers[0] && res.transfers[0].recipient);
        if (recipient) {
            res.sender = (0, ts_lib_crypto_1.address)({ publicKey: t.senderPublicKey }, (0, generic_1.chainIdFromRecipient)(recipient));
        }
    }
    return res;
}
function orderToProtoBytes(obj) {
    return wavesProto.waves.Order.encode(orderToProto(obj)).finish();
}
function protoBytesToOrder(bytes) {
    const o = wavesProto.waves.Order.decode(bytes);
    return orderFromProto(o);
}
const getCommonFields = ({ senderPublicKey, fee, timestamp, type, version, ...rest }) => {
    const typename = nameByType[type];
    let chainId = rest.chainId;
    if (chainId == null) {
        const r = rest;
        let recipient = r.recipient || r.dApp || (r.transfers && r.transfers[0] && r.transfers[0].recipient);
        if (recipient) {
            chainId = (0, generic_1.chainIdFromRecipient)(recipient);
        }
    }
    return {
        version,
        type,
        chainId,
        senderPublicKey: (0, ts_lib_crypto_1.base58Decode)(senderPublicKey),
        timestamp: long_1.default.fromValue(timestamp),
        fee: amountToProto(fee, rest.feeAssetId),
        data: typename,
    };
};
const getCommonSignedFields = (tx) => {
    const fields = getCommonFields(tx);
    if (Object.hasOwn(tx, 'proofs')) {
        fields.proofs = tx.proofs;
    }
    return fields;
};
const getIssueData = (t) => ({
    name: t.name,
    description: t.description === '' ? null : t.description,
    amount: long_1.default.fromValue(t.quantity),
    decimals: t.decimals === 0 ? null : t.decimals,
    reissuable: t.reissuable ? true : undefined,
    script: t.script == null ? null : (0, exports.scriptToProto)(t.script),
});
const getTransferData = (t) => ({
    recipient: recipientToProto(t.recipient),
    amount: amountToProto(t.amount, t.assetId),
    attachment: t.attachment == null || t.attachment == '' ? undefined : (0, ts_lib_crypto_1.base58Decode)(t.attachment),
});
const getReissueData = (t) => ({
    assetAmount: amountToProto(t.quantity, t.assetId),
    reissuable: t.reissuable ? true : undefined,
});
const getBurnData = (t) => ({
    assetAmount: amountToProto(t.amount || t.amount, t.assetId),
});
const getExchangeData = (t) => ({
    amount: long_1.default.fromValue(t.amount),
    price: long_1.default.fromValue(t.price),
    buyMatcherFee: long_1.default.fromValue(t.buyMatcherFee),
    sellMatcherFee: long_1.default.fromValue(t.sellMatcherFee),
    orders: [orderToProto({ chainId: t.chainId, ...t.order1 }), orderToProto({ chainId: t.chainId, ...t.order2 })],
});
const getLeaseData = (t) => ({
    recipient: recipientToProto(t.recipient),
    amount: long_1.default.fromValue(t.amount),
});
const getCancelLeaseData = (t) => ({
    leaseId: (0, ts_lib_crypto_1.base58Decode)(t.leaseId),
});
const getAliasData = (t) => ({ alias: t.alias });
const getMassTransferData = (t) => ({
    assetId: t.assetId == null ? null : (0, ts_lib_crypto_1.base58Decode)(t.assetId),
    attachment: t.attachment == null || t.attachment == '' ? undefined : (0, ts_lib_crypto_1.base58Decode)(t.attachment),
    transfers: t.transfers.map(massTransferItemToProto),
});
const getDataTxData = (t) => ({
    data: t.data.map(exports.dataEntryToProto),
});
const getSetScriptData = (t) => ({
    script: t.script == null ? null : (0, exports.scriptToProto)(t.script),
});
const getSponsorData = (t) => ({
    minFee: t.minSponsoredAssetFee === null ? amountToProto(0, t.assetId) : amountToProto(t.minSponsoredAssetFee, t.assetId),
});
const getSetAssetScriptData = (t) => ({
    assetId: (0, ts_lib_crypto_1.base58Decode)(t.assetId),
    script: t.script == null ? null : (0, exports.scriptToProto)(t.script),
});
const getInvokeData = (t) => ({
    dApp: recipientToProto(t.dApp),
    functionCall: marshall_1.binary.serializerFromSchema(marshall_1.schemas.invokeScriptSchemaV1.schema[5][1])(t.call),
    payments: t.payment == null ? null : t.payment.map(({ amount, assetId }) => amountToProto(amount, assetId)),
});
const getUpdateAssetInfoData = (t) => {
    return {
        assetId: (0, ts_lib_crypto_1.base58Decode)(t.assetId),
        name: t.name,
        description: t.description === '' ? null : t.description,
    };
};
const getCommitToGenerationData = (t) => {
    const tx = t;
    return {
        generationPeriodStart: t.generationPeriodStart,
        endorserPublicKey: tx.endorserPublicKey == null ? undefined : (0, ts_lib_crypto_1.base58Decode)(tx.endorserPublicKey),
        commitmentSignature: tx.commitmentSignature == null ? undefined : (0, ts_lib_crypto_1.base58Decode)(tx.commitmentSignature),
    };
};
// const getInvokeExpressionData = (t: InvokeExpressionTransaction): wavesProto.waves.IInvokeExpressionTransactionData => {
//     return {
//         expression: t.expression == null ? null : scriptToProto(t.expression),
//     }
// }
const getTxData = (t) => {
    let txData;
    switch (t.type) {
        case ts_types_1.TRANSACTION_TYPE.ISSUE:
            txData = getIssueData(t);
            break;
        case ts_types_1.TRANSACTION_TYPE.TRANSFER:
            txData = getTransferData(t);
            break;
        case ts_types_1.TRANSACTION_TYPE.REISSUE:
            txData = getReissueData(t);
            break;
        case ts_types_1.TRANSACTION_TYPE.BURN:
            txData = getBurnData(t);
            break;
        case ts_types_1.TRANSACTION_TYPE.LEASE:
            txData = getLeaseData(t);
            break;
        case ts_types_1.TRANSACTION_TYPE.CANCEL_LEASE:
            txData = getCancelLeaseData(t);
            break;
        case ts_types_1.TRANSACTION_TYPE.ALIAS:
            txData = getAliasData(t);
            break;
        case ts_types_1.TRANSACTION_TYPE.MASS_TRANSFER:
            txData = getMassTransferData(t);
            break;
        case ts_types_1.TRANSACTION_TYPE.DATA:
            txData = getDataTxData(t);
            break;
        case ts_types_1.TRANSACTION_TYPE.SET_SCRIPT:
            txData = getSetScriptData(t);
            break;
        case ts_types_1.TRANSACTION_TYPE.SET_ASSET_SCRIPT:
            txData = getSetAssetScriptData(t);
            break;
        case ts_types_1.TRANSACTION_TYPE.SPONSORSHIP:
            txData = getSponsorData(t);
            break;
        case ts_types_1.TRANSACTION_TYPE.EXCHANGE:
            txData = getExchangeData(t);
            break;
        case ts_types_1.TRANSACTION_TYPE.INVOKE_SCRIPT:
            txData = getInvokeData(t);
            break;
        case ts_types_1.TRANSACTION_TYPE.UPDATE_ASSET_INFO:
            txData = getUpdateAssetInfoData(t);
            break;
        case ts_types_1.TRANSACTION_TYPE.COMMIT_TO_GENERATION:
            txData = getCommitToGenerationData(t);
            break;
        // case TRANSACTION_TYPE.INVOKE_EXPRESSION:
        //     txData = getInvokeExpressionData(t)
        //     break
    }
    return txData;
};
const txToProto = (t) => {
    const common = getCommonFields(t);
    const txData = getTxData(t);
    return {
        ...common,
        [common.data]: txData,
    };
};
exports.txToProto = txToProto;
const signedTxToProto = (t) => {
    const common = getCommonSignedFields(t);
    const txData = getTxData(t);
    return {
        wavesTransaction: {
            ...common,
            [common.data]: txData,
        },
        proofs: (t.proofs || []).map(proof2Uint8Array),
    };
};
exports.signedTxToProto = signedTxToProto;
const orderToProto = (o) => {
    let priceMode;
    let attachment;
    if (o.version === 4 && 'priceMode' in o) {
        if (o.priceMode === 0 || o.priceMode === 'default') {
            priceMode = undefined;
        }
        else {
            priceMode = o.priceMode === 'assetDecimals'
                ? wavesProto.waves.Order.PriceMode.ASSET_DECIMALS
                : wavesProto.waves.Order.PriceMode.FIXED_DECIMALS;
        }
    }
    else
        priceMode = undefined;
    if (o.version === 4 && 'attachment' in o) {
        attachment = (0, ts_lib_crypto_1.base64Decode)(o.attachment.startsWith('base64:') ? o.attachment.slice(7) : o.attachment);
    }
    else
        attachment = undefined;
    const isNullOrWaves = (asset) => asset == null || asset.toLowerCase() == 'waves';
    return ({
        chainId: o.chainId,
        senderPublicKey: o.senderPublicKey ? (0, ts_lib_crypto_1.base58Decode)(o.senderPublicKey) : null,
        matcherPublicKey: (0, ts_lib_crypto_1.base58Decode)(o.matcherPublicKey),
        assetPair: {
            amountAssetId: isNullOrWaves(o.assetPair.amountAsset) ? null : (0, ts_lib_crypto_1.base58Decode)(o.assetPair.amountAsset),
            priceAssetId: isNullOrWaves(o.assetPair.priceAsset) ? null : (0, ts_lib_crypto_1.base58Decode)(o.assetPair.priceAsset),
        },
        orderSide: o.orderType === 'buy' ? undefined : wavesProto.waves.Order.Side.SELL,
        amount: long_1.default.fromValue(o.amount),
        price: long_1.default.fromValue(o.price),
        timestamp: long_1.default.fromValue(o.timestamp),
        expiration: long_1.default.fromValue(o.expiration),
        matcherFee: amountToProto(o.matcherFee, o.matcherFeeAssetId ? o.matcherFeeAssetId : null),
        version: o.version,
        proofs: o.proofs?.map(ts_lib_crypto_1.base58Decode),
        eip712Signature: o.eip712Signature ? (0, ts_lib_crypto_1.base16Decode)(o.eip712Signature.slice(2)) : undefined,
        priceMode: priceMode,
        attachment: attachment,
    });
};
const orderFromProto = (po) => {
    let priceMode;
    if (po.version === 4 && po.priceMode) {
        po.priceMode === 1
            ? priceMode = 'fixedDecimals'
            : priceMode = 'assetDecimals';
    }
    else
        priceMode = undefined;
    return {
        version: po.version,
        senderPublicKey: (0, ts_lib_crypto_1.base58Encode)(po.senderPublicKey),
        matcherPublicKey: (0, ts_lib_crypto_1.base58Encode)(po.matcherPublicKey),
        assetPair: {
            amountAsset: po.assetPair.amountAssetId == null ? null : (0, ts_lib_crypto_1.base58Encode)(po.assetPair.amountAssetId),
            priceAsset: po.assetPair.priceAssetId == null ? null : (0, ts_lib_crypto_1.base58Encode)(po.assetPair.priceAssetId),
        },
        // @ts-ignore
        chainId: po.chainId,
        orderType: po.orderSide === wavesProto.waves.Order.Side.BUY ? 'buy' : 'sell',
        amount: convertNumber(po.amount),
        price: convertNumber(po.price),
        timestamp: po.timestamp.toNumber(),
        expiration: po.expiration.toNumber(),
        matcherFee: convertNumber(po.matcherFee.amount),
        matcherFeeAssetId: po.matcherFee.assetId == null ? null : (0, ts_lib_crypto_1.base58Encode)(po.matcherFee.assetId),
        attachment: po.attachment?.length ? (0, ts_lib_crypto_1.base64Encode)(po.attachment) : undefined,
        // @ts-ignore
        priceMode: priceMode,
        eip712Signature: po.eip712Signature?.length ? `0x${(0, ts_lib_crypto_1.base16Encode)(po.eip712Signature)}` : undefined,
    };
};
const recipientToProto = (r) => ({
    alias: r.startsWith('alias') ? r.slice(8) : undefined,
    publicKeyHash: !r.startsWith('alias') ? (0, ts_lib_crypto_1.base58Decode)(r).slice(2, -4) : undefined,
});
const amountToProto = (a, assetId) => ({
    amount: a == 0 ? null : long_1.default.fromValue(a),
    assetId: assetId == null ? null : (0, ts_lib_crypto_1.base58Decode)(assetId),
});
const massTransferItemToProto = (mti) => ({
    recipient: recipientToProto(mti.recipient),
    amount: mti.amount == 0 ? null : long_1.default.fromValue(mti.amount),
});
const dataEntryToProto = (de) => ({
    key: de.key,
    intValue: de.type === 'integer' ? long_1.default.fromValue(de.value) : undefined,
    boolValue: de.type === 'boolean' ? de.value : undefined,
    binaryValue: de.type === 'binary' ? (0, ts_lib_crypto_1.base64Decode)((de.value.startsWith('base64:') ? de.value.slice(7) : de.value)) : undefined,
    stringValue: de.type === 'string' ? de.value : undefined,
});
exports.dataEntryToProto = dataEntryToProto;
const scriptToProto = (s) => {
    return s ? (0, ts_lib_crypto_1.base64Decode)(s.toString().startsWith('base64:') ? s.slice(7) : s) : null;
};
exports.scriptToProto = scriptToProto;
const nameByType = {
    1: 'genesis',
    2: 'payment',
    3: 'issue',
    4: 'transfer',
    5: 'reissue',
    6: 'burn',
    7: 'exchange',
    8: 'lease',
    9: 'leaseCancel',
    10: 'createAlias',
    11: 'massTransfer',
    12: 'dataTransaction',
    13: 'setScript',
    14: 'sponsorFee',
    15: 'setAssetScript',
    16: 'invokeScript',
    17: 'updateAssetInfo',
    19: 'commitToGeneration',
};
const typeByName = {
    'genesis': 1,
    'payment': 2,
    'issue': 3,
    'transfer': 4,
    'reissue': 5,
    'burn': 6,
    'exchange': 7,
    'lease': 8,
    'leaseCancel': 9,
    'createAlias': 10,
    'massTransfer': 11,
    'dataTransaction': 12,
    'setScript': 13,
    'sponsorFee': 14,
    'setAssetScript': 15,
    'invokeScript': 16,
    'updateAssetInfo': 17,
    'commitToGeneration': 19,
};
const proof2Uint8Array = (proof) => {
    return (0, ts_lib_crypto_1.base58Decode)(proof);
};
const uint8Array2proof = (proofBytes) => {
    return (0, ts_lib_crypto_1.base58Encode)(proofBytes);
};
