// @waves/waves-transactions/dist/validators/data.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.dataValidator = exports.dataFieldValidator = void 0;
const ts_types_1 = require("@waves/ts-types");
const validators_1 = require("./validators");
const dataScheme = {
    type: (0, validators_1.isEq)(ts_types_1.TRANSACTION_TYPE.DATA),
    senderPublicKey: validators_1.isPublicKey,
    version: (0, validators_1.orEq)([undefined, 1, 2]),
    data: (data) => (0, validators_1.isArray)(data) &&
        data.every(item => (0, validators_1.isValidData)(item) || (0, validators_1.isValidDeleteRequest)(item)),
    fee: validators_1.isNaturalNumberOrZeroLike,
    timestamp: validators_1.isNumber,
    proofs: (0, validators_1.ifElse)(validators_1.isArray, (0, validators_1.defaultValue)(true), (0, validators_1.orEq)([undefined])),
};
const dataFieldValidator = (item) => (0, validators_1.isValidData)(item) || (0, validators_1.isValidDeleteRequest)(item);
exports.dataFieldValidator = dataFieldValidator;
exports.dataValidator = (0, validators_1.validateByShema)(dataScheme, validators_1.getError);
