// @waves/waves-transactions/dist/validators/exchange.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.exchangeValidator = void 0;
const ts_types_1 = require("@waves/ts-types");
const order_1 = require("./order");
const validators_1 = require("./validators");
const exchangeScheme = {
    type: (0, validators_1.isEq)(ts_types_1.TRANSACTION_TYPE.EXCHANGE),
    senderPublicKey: validators_1.isPublicKey,
    version: (0, validators_1.orEq)([undefined, 1, 2, 3]),
    order1: (0, validators_1.validatePipe)((0, validators_1.isRequired)(true), order_1.orderValidator),
    order2: (0, validators_1.validatePipe)((0, validators_1.isRequired)(true), order_1.orderValidator),
    amount: validators_1.isNumberLike,
    price: validators_1.isNumberLike,
    buyMatcherFee: validators_1.isNumberLike,
    sellMatcherFee: validators_1.isNumberLike,
    fee: validators_1.isNaturalNumberOrZeroLike,
    timestamp: validators_1.isNumber,
    proofs: (0, validators_1.ifElse)(validators_1.isArray, (0, validators_1.defaultValue)(true), (0, validators_1.orEq)([undefined])),
};
exports.exchangeValidator = (0, validators_1.validateByShema)(exchangeScheme, validators_1.getError);
