// @waves/waves-transactions/dist/validators/invoke-script.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.invokeValidator = void 0;
const ts_types_1 = require("@waves/ts-types");
const validators_1 = require("./validators");
const invokeScheme = {
    type: (0, validators_1.isEq)(ts_types_1.TRANSACTION_TYPE.INVOKE_SCRIPT),
    senderPublicKey: validators_1.isPublicKey,
    version: (0, validators_1.orEq)([undefined, 1, 2]),
    dApp: validators_1.isRecipient,
    call: (0, validators_1.ifElse)((0, validators_1.isRequired)(false), (0, validators_1.defaultValue)(true), (0, validators_1.validatePipe)((0, validators_1.pipe)((0, validators_1.prop)('function'), validators_1.isString), (0, validators_1.pipe)((0, validators_1.prop)('function'), (0, validators_1.prop)('length'), (0, validators_1.gte)(0)), (0, validators_1.pipe)((0, validators_1.prop)('args'), validators_1.isArray), (data) => data.every((0, validators_1.validatePipe)((0, validators_1.isRequired)(true), validators_1.isValidDataPair)))),
    payment: (0, validators_1.validatePipe)(validators_1.isArray, (data) => data.every((0, validators_1.validatePipe)((0, validators_1.pipe)((0, validators_1.prop)('amount'), validators_1.isNumberLike), (0, validators_1.pipe)((0, validators_1.prop)('assetId'), validators_1.isWavesOrAssetId)))),
    fee: validators_1.isNaturalNumberOrZeroLike,
    feeAssetId: validators_1.isWavesOrAssetId,
    chainId: validators_1.isNaturalNumberLike,
    timestamp: validators_1.isNaturalNumberLike,
    proofs: (0, validators_1.ifElse)(validators_1.isArray, (0, validators_1.defaultValue)(true), (0, validators_1.orEq)([undefined])),
};
exports.invokeValidator = (0, validators_1.validateByShema)(invokeScheme, validators_1.getError);
