// @waves/waves-transactions/dist/validators/issue.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.issueValidator = void 0;
const ts_types_1 = require("@waves/ts-types");
const validators_1 = require("./validators");
const issueScheme = {
    type: (0, validators_1.isEq)(ts_types_1.TRANSACTION_TYPE.ISSUE),
    version: (0, validators_1.orEq)([undefined, 2, 3]),
    senderPublicKey: validators_1.isPublicKey,
    name: validators_1.isValidAssetName,
    description: validators_1.isValidAssetDescription,
    quantity: validators_1.isNaturalNumberLike,
    decimals: validators_1.isNumber,
    reissuable: validators_1.isBoolean,
    script: (0, validators_1.ifElse)((0, validators_1.isRequired)(true), validators_1.isBase64, (0, validators_1.defaultValue)(true)),
    chainId: validators_1.isNumber,
    fee: validators_1.isNaturalNumberOrZeroLike,
    timestamp: validators_1.isNumber,
    proofs: (0, validators_1.ifElse)(validators_1.isArray, (0, validators_1.defaultValue)(true), (0, validators_1.orEq)([undefined])),
};
exports.issueValidator = (0, validators_1.validateByShema)(issueScheme, validators_1.getError);
