// @waves/waves-transactions/dist/validators/mass-transfer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.massTransferValidator = void 0;
const ts_types_1 = require("@waves/ts-types");
const validators_1 = require("./validators");
const massTransferScheme = {
    type: (0, validators_1.isEq)(ts_types_1.TRANSACTION_TYPE.MASS_TRANSFER),
    senderPublicKey: validators_1.isPublicKey,
    version: (0, validators_1.orEq)([undefined, 1, 2]),
    transfers: (0, validators_1.validatePipe)(validators_1.isArray, (0, validators_1.pipe)((0, validators_1.prop)('length'), (0, validators_1.gte)(0)), (0, validators_1.pipe)((0, validators_1.prop)('length'), (0, validators_1.lte)(100)), (data) => data.every((0, validators_1.validatePipe)((0, validators_1.isRequired)(true), (0, validators_1.pipe)((0, validators_1.prop)('recipient'), validators_1.isRecipient), (0, validators_1.pipe)((0, validators_1.prop)('amount'), validators_1.isNaturalNumberOrZeroLike)))),
    assetId: validators_1.isWavesOrAssetId,
    attachment: validators_1.isAttachment,
    fee: validators_1.isNaturalNumberOrZeroLike,
    timestamp: validators_1.isNumber,
    proofs: (0, validators_1.ifElse)(validators_1.isArray, (0, validators_1.defaultValue)(true), (0, validators_1.orEq)([undefined])),
};
exports.massTransferValidator = (0, validators_1.validateByShema)(massTransferScheme, validators_1.getError);
