// @waves/waves-transactions/dist/validators/order.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.orderValidator = void 0;
const validators_1 = require("./validators");
const orderScheme = {
    orderType: (0, validators_1.orEq)(['sell', 'buy']),
    matcherPublicKey: validators_1.isPublicKey,
    version: (0, validators_1.orEq)([undefined, 1, 2, 3, 4]),
    assetPair: (0, validators_1.validatePipe)((0, validators_1.isRequired)(true), (0, validators_1.pipe)((0, validators_1.prop)('amountAsset'), validators_1.isWavesOrAssetId), (0, validators_1.pipe)((0, validators_1.prop)('priceAsset'), validators_1.isWavesOrAssetId)),
    price: validators_1.isNumberLike,
    amount: validators_1.isNumberLike,
    matcherFee: validators_1.isNumberLike,
    expiration: validators_1.isNumberLike,
    timestamp: validators_1.isNumber,
    proofs: (0, validators_1.ifElse)(validators_1.isArray, (0, validators_1.defaultValue)(true), (0, validators_1.orEq)([undefined])),
};
const v1_2_OrderScheme = {
    matcherFeeAssetId: (0, validators_1.orEq)([undefined, null, 'WAVES']),
    senderPublicKey: validators_1.isPublicKey,
};
const v3_OrderScheme = {
    matcherFeeAssetId: validators_1.isWavesOrAssetId,
    senderPublicKey: validators_1.isPublicKey,
};
const v4_OrderScheme = {
    matcherFeeAssetId: validators_1.isWavesOrAssetId,
    senderPublicKey: validators_1.isPublicKeyForEthSuppTx,
};
const validateOrder = (0, validators_1.validateByShema)(orderScheme, validators_1.getError);
const validateOrderV2 = (0, validators_1.validateByShema)(v1_2_OrderScheme, validators_1.getError);
const validateOrderV3 = (0, validators_1.validateByShema)(v3_OrderScheme, validators_1.getError);
const validateOrderV4 = (0, validators_1.validateByShema)(v4_OrderScheme, validators_1.getError);
exports.orderValidator = (0, validators_1.validatePipe)(validateOrder, (0, validators_1.ifElse)((0, validators_1.pipe)((0, validators_1.prop)('version'), (v) => v >= 3), (0, validators_1.ifElse)((0, validators_1.pipe)((0, validators_1.prop)('version'), (v) => v >= 4), validateOrderV4, validateOrderV3), validateOrderV2));
