// @waves/waves-transactions/dist/validators/validators.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getError = exports.validateByShema = exports.isRecipient = exports.exception = exports.isValidAssetDescription = exports.isValidAssetName = exports.isValidDeleteRequest = exports.isValidData = exports.isValidDataPair = exports.isAttachment = exports.isAssetId = exports.isWavesOrAssetId = exports.isPublicKeyForEthSuppTx = exports.isPublicKey = exports.isHash = exports.isValidAlias = exports.isValidAliasName = exports.isValidAddress = exports.isBase64 = exports.isBase58 = exports.bytesLength = exports.isArray = exports.isByteArray = exports.isBoolean = exports.isNaturalNumberOrNullLike = exports.isNaturalNumberOrZeroLike = exports.isNaturalNumberLike = exports.isNumberLike = exports.isNumber = exports.isString = exports.isRequired = exports.orEq = exports.isEq = exports.ifElse = exports.gte = exports.lte = exports.prop = exports.validatePipe = exports.pipe = exports.nope = exports.defaultValue = void 0;
const ts_lib_crypto_1 = require("@waves/ts-lib-crypto");
const TX_DEFAULTS = {
    MAX_ATTACHMENT: 140,
    ALIAS: {
        AVAILABLE_CHARS: '-.0123456789@_abcdefghijklmnopqrstuvwxyz',
        MAX_ALIAS_LENGTH: 30,
        MIN_ALIAS_LENGTH: 4,
    },
};
const ASSETS = {
    NAME_MIN_BYTES: 4,
    NAME_MAX_BYTES: 16,
    DESCRIPTION_MAX_BYTES: 1000,
};
const defaultValue = (value) => () => value;
exports.defaultValue = defaultValue;
const nope = (value) => value;
exports.nope = nope;
const pipe = (...args) => (value) => args.reduce((acc, cb) => cb(acc), value);
exports.pipe = pipe;
const validatePipe = (...args) => (value) => {
    let isValid = true;
    for (const cb of args) {
        isValid = !!cb(value);
        if (!isValid) {
            return false;
        }
    }
    return isValid;
};
exports.validatePipe = validatePipe;
const prop = (key) => (value) => value ? value[key] : undefined;
exports.prop = prop;
const lte = (ref) => (value) => ref >= value;
exports.lte = lte;
const gte = (ref) => (value) => ref <= value;
exports.gte = gte;
const ifElse = (condition, a, b) => (value) => condition(value) ? a(value) : b(value);
exports.ifElse = ifElse;
const isEq = (reference) => (value) => {
    switch (true) {
        case (0, exports.isNumber)(value) && (0, exports.isNumber)(reference):
            return Number(value) === Number(reference);
        case (0, exports.isString)(value) && (0, exports.isString)(reference):
            return String(reference) === String(value);
        case (0, exports.isBoolean)(value) && (0, exports.isBoolean)(reference):
            return Boolean(value) === Boolean(reference);
        default:
            return reference === value;
    }
};
exports.isEq = isEq;
const orEq = (referencesList) => (value) => referencesList.some((0, exports.isEq)(value));
exports.orEq = orEq;
const isRequired = (required) => (value) => !required || value != null;
exports.isRequired = isRequired;
const isString = (value) => typeof value === 'string' || value instanceof String;
exports.isString = isString;
const isNumber = (value) => (typeof value === 'number' || value instanceof Number) && !isNaN(Number(value));
exports.isNumber = isNumber;
const isNumberLike = (value) => value != null && !isNaN(Number(value)) && !!(value || value === 0);
exports.isNumberLike = isNumberLike;
const isNaturalNumberLike = (value) => value != null && !isNaN(Number(value)) && Number(value) > 0;
exports.isNaturalNumberLike = isNaturalNumberLike;
const isNaturalNumberOrZeroLike = (value) => value != null && !isNaN(Number(value)) && Number(value) >= 0;
exports.isNaturalNumberOrZeroLike = isNaturalNumberOrZeroLike;
const isNaturalNumberOrNullLike = (value) => (!isNaN(Number(value)) && Number(value) > 0) || value === null;
exports.isNaturalNumberOrNullLike = isNaturalNumberOrNullLike;
const isBoolean = (value) => value != null && (typeof value === 'boolean' || value instanceof Boolean);
exports.isBoolean = isBoolean;
const isByteArray = (value) => {
    if (!value) {
        return false;
    }
    const bytes = new Uint8Array(value);
    return bytes.length === value.length && bytes.every((val, index) => (0, exports.isEq)(val)(value[index]));
};
exports.isByteArray = isByteArray;
const isArray = (value) => Array.isArray(value);
exports.isArray = isArray;
const bytesLength = (length) => (value) => {
    try {
        return Uint8Array.from(value).length === length;
    }
    catch (e) {
        return false;
    }
};
exports.bytesLength = bytesLength;
const isBase58 = (value) => {
    try {
        (0, ts_lib_crypto_1.base58Decode)(value);
    }
    catch (e) {
        return false;
    }
    return true;
};
exports.isBase58 = isBase58;
const isBase64 = (value) => {
    try {
        value = value.replace(/^base64:/, '');
        (0, ts_lib_crypto_1.base64Decode)(value);
    }
    catch (e) {
        return false;
    }
    return true;
};
exports.isBase64 = isBase64;
const isValidAddress = (address, network) => {
    if (typeof address !== 'string' || !(0, exports.isBase58)(address)) {
        return false;
    }
    let addressBytes = (0, ts_lib_crypto_1.base58Decode)(address);
    if (addressBytes[0] !== 1) {
        return false;
    }
    if (network != null && addressBytes[1] != network) {
        return false;
    }
    let key = addressBytes.slice(0, 22);
    let check = addressBytes.slice(22, 26);
    let keyHash = (0, ts_lib_crypto_1.keccak)((0, ts_lib_crypto_1.blake2b)(key)).slice(0, 4);
    for (let i = 0; i < 4; i++) {
        if (check[i] !== keyHash[i]) {
            return false;
        }
    }
    return true;
};
exports.isValidAddress = isValidAddress;
const validateChars = (chars) => (value) => value.split('').every((char) => chars.includes(char));
exports.isValidAliasName = (0, exports.ifElse)(validateChars(TX_DEFAULTS.ALIAS.AVAILABLE_CHARS), (0, exports.pipe)((0, exports.prop)('length'), (0, exports.validatePipe)((0, exports.lte)(TX_DEFAULTS.ALIAS.MAX_ALIAS_LENGTH), (0, exports.gte)(TX_DEFAULTS.ALIAS.MIN_ALIAS_LENGTH))), (0, exports.defaultValue)(false));
exports.isValidAlias = (0, exports.validatePipe)(exports.isString, (0, exports.pipe)((value) => value.split(':'), (0, exports.ifElse)((value) => value[0] !== 'alias' || value.length !== 3, (0, exports.defaultValue)(false), (0, exports.pipe)((0, exports.prop)(2), exports.isValidAliasName))));
exports.isHash = (0, exports.validatePipe)((0, exports.isRequired)(true), exports.isBase58, (0, exports.pipe)((value) => (0, ts_lib_crypto_1.base58Decode)(value), (0, exports.bytesLength)(32)));
exports.isPublicKey = exports.isHash;
exports.isPublicKeyForEthSuppTx = (0, exports.ifElse)((0, exports.orEq)(['', null, undefined]), (0, exports.defaultValue)(true), (0, exports.pipe)((value) => (0, ts_lib_crypto_1.base58Decode)(value), (value) => {
    try {
        return Uint8Array.from(value).length === 32 || Uint8Array.from(value).length === 64;
    }
    catch (e) {
        return false;
    }
}));
exports.isWavesOrAssetId = (0, exports.ifElse)((0, exports.orEq)(['', null, undefined, 'WAVES']), (0, exports.defaultValue)(true), exports.isHash);
exports.isAssetId = exports.isHash;
exports.isAttachment = (0, exports.ifElse)((0, exports.orEq)([null, undefined]), (0, exports.defaultValue)(true), (0, exports.ifElse)(
// if valid Data Pair
(0, exports.validatePipe)(exports.isArray, (data) => data.every(exports.isValidDataPair)), (0, exports.defaultValue)(true), 
// else if valid base58 or bytearray
(0, exports.pipe)((0, exports.ifElse)(exports.isBase58, ts_lib_crypto_1.base58Decode, exports.nope), (0, exports.ifElse)(exports.isByteArray, (0, exports.pipe)((0, exports.prop)('length'), (0, exports.lte)(TX_DEFAULTS.MAX_ATTACHMENT)), (0, exports.defaultValue)(false)))));
const validateType = {
    integer: exports.isNumberLike,
    boolean: exports.isBoolean,
    string: exports.isString,
    binary: exports.isBase64,
    list: exports.isArray,
};
const isValidDataPair = (data) => !!(validateType[data.type] && validateType[data.type](data.value));
exports.isValidDataPair = isValidDataPair;
exports.isValidData = (0, exports.validatePipe)((0, exports.isRequired)(true), (0, exports.pipe)((0, exports.prop)('key'), (0, exports.validatePipe)(exports.isString, (key) => !!key)), exports.isValidDataPair);
exports.isValidDeleteRequest = (0, exports.validatePipe)((0, exports.isRequired)(true), (0, exports.pipe)((0, exports.prop)('key'), (0, exports.validatePipe)(exports.isString, (key) => !!key)), ({ type, value }) => type == null && value == null);
exports.isValidAssetName = (0, exports.validatePipe)((0, exports.isRequired)(true), exports.isString, (0, exports.pipe)(ts_lib_crypto_1.stringToBytes, (0, exports.prop)('length'), (0, exports.ifElse)((0, exports.gte)(ASSETS.NAME_MIN_BYTES), (0, exports.lte)(ASSETS.NAME_MAX_BYTES), (0, exports.defaultValue)(false))));
exports.isValidAssetDescription = (0, exports.validatePipe)((0, exports.isRequired)(false), (0, exports.defaultValue)(true), (0, exports.pipe)(ts_lib_crypto_1.stringToBytes, (0, exports.prop)('length'), (0, exports.lte)(ASSETS.DESCRIPTION_MAX_BYTES)));
const exception = (msg) => {
    throw new Error(msg);
};
exports.exception = exception;
exports.isRecipient = (0, exports.ifElse)(exports.isValidAddress, (0, exports.defaultValue)(true), exports.isValidAlias);
const validateByShema = (shema, errorTpl) => (tx) => {
    Object.entries(shema).forEach(([key, cb]) => {
        const value = (0, exports.prop)(key)(tx || {});
        if (!cb(value)) {
            (0, exports.exception)(errorTpl(key, value));
        }
    });
    return true;
};
exports.validateByShema = validateByShema;
const getError = (key, value) => {
    return `tx "${key}", has wrong data: ${JSON.stringify(value)}. Check tx data.`;
};
exports.getError = getError;
