// @waves/waves-transactions/dist/transactions/alias.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.alias = alias;
/**
 * @module index
 */
const marshall_1 = require("@waves/marshall");
const ts_lib_crypto_1 = require("@waves/ts-lib-crypto");
const ts_types_1 = require("@waves/ts-types");
const proto_serialize_1 = require("../proto-serialize");
const generic_1 = require("../generic");
const validators_1 = require("../validators");
const defaultVersions_1 = require("../defaultVersions");
function alias(paramsOrTx, seed) {
    const type = ts_types_1.TRANSACTION_TYPE.ALIAS;
    const version = paramsOrTx.version || defaultVersions_1.DEFAULT_VERSIONS.ALIAS;
    const seedsAndIndexes = (0, generic_1.convertToPairs)(seed);
    const senderPublicKey = (0, generic_1.getSenderPublicKey)(seedsAndIndexes, paramsOrTx);
    const tx = {
        type,
        version,
        senderPublicKey,
        alias: paramsOrTx.alias,
        fee: (0, generic_1.fee)(paramsOrTx, 100000),
        timestamp: paramsOrTx.timestamp || Date.now(),
        chainId: (0, generic_1.networkByte)(paramsOrTx.chainId, 87),
        proofs: paramsOrTx.proofs || [],
        id: '',
    };
    validators_1.validate.alias(tx);
    const bytes = version > 2 ? (0, proto_serialize_1.txToProtoBytes)(tx) : marshall_1.binary.serializeTx(tx);
    const idBytes = version > 2 ? bytes : [bytes[0], ...bytes.slice(36, -16)];
    seedsAndIndexes.forEach(([s, i]) => (0, generic_1.addProof)(tx, (0, ts_lib_crypto_1.signBytes)(s, bytes), i));
    tx.id = (0, ts_lib_crypto_1.base58Encode)((0, ts_lib_crypto_1.blake2b)(Uint8Array.from(idBytes)));
    return tx;
}
