// @waves/waves-transactions/dist/transactions/commit-to-generation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.commitToGeneration = commitToGeneration;
/**
 * @module index
 */
const ts_lib_crypto_1 = require("@waves/ts-lib-crypto");
const ts_types_1 = require("@waves/ts-types");
const generic_1 = require("../generic");
const validators_1 = require("../validators");
const proto_serialize_1 = require("../proto-serialize");
const defaultVersions_1 = require("../defaultVersions");
const wavesCrypto = (0, ts_lib_crypto_1.crypto)({ output: 'Bytes' });
const int32ToBigEndianBytes = (value) => {
    if (!Number.isInteger(value) || value < -2147483648 || value > 2147483647) {
        throw new Error(`generationPeriodStart should be a 32-bit integer, but got: ${value}`);
    }
    const result = new Uint8Array(4);
    new DataView(result.buffer).setInt32(0, value, false);
    return result;
};
function commitToGeneration(paramsOrTx, seed) {
    const type = ts_types_1.TRANSACTION_TYPE.COMMIT_TO_GENERATION;
    const version = paramsOrTx.version || defaultVersions_1.DEFAULT_VERSIONS.COMMIT_TO_GENERATION;
    const seedsAndIndexes = (0, generic_1.convertToPairs)(seed);
    const senderPublicKey = (0, generic_1.getSenderPublicKey)(seedsAndIndexes, paramsOrTx);
    const primarySeed = seedsAndIndexes[0]?.[0];
    const shouldComputeBls = paramsOrTx.endorserPublicKey == null || paramsOrTx.commitmentSignature == null;
    const blsKeyPair = !shouldComputeBls || primarySeed == null
        ? undefined
        : wavesCrypto.blsKeyPair((0, ts_lib_crypto_1.isPrivateKey)(primarySeed)
            ? primarySeed.privateKey
            : (0, ts_lib_crypto_1.privateKey)(primarySeed));
    const blsSecret = blsKeyPair?.blsSecret;
    const endorserPublicKey = paramsOrTx.endorserPublicKey == null
        ? blsKeyPair == null
            ? undefined
            : (0, ts_lib_crypto_1.base58Encode)(blsKeyPair.blsPublic)
        : paramsOrTx.endorserPublicKey;
    if (endorserPublicKey == null) {
        throw new Error('Please provide either seed or endorserPublicKey for CommitToGeneractionTransaction');
    }
    const commitmentSignature = paramsOrTx.commitmentSignature == null
        ? blsSecret == null
            ? undefined
            : (0, ts_lib_crypto_1.base58Encode)(wavesCrypto.blsSign(blsSecret, (0, ts_lib_crypto_1.concat)((0, ts_lib_crypto_1.base58Decode)(endorserPublicKey), int32ToBigEndianBytes(paramsOrTx.generationPeriodStart))))
        : paramsOrTx.commitmentSignature;
    if (commitmentSignature == null) {
        throw new Error('Please provide either seed or commitmentSignature for CommitToGeneractionTransaction');
    }
    const tx = {
        type,
        version,
        senderPublicKey,
        generationPeriodStart: paramsOrTx.generationPeriodStart,
        endorserPublicKey,
        commitmentSignature,
        fee: (0, generic_1.fee)(paramsOrTx, 10000000),
        timestamp: paramsOrTx.timestamp || Date.now(),
        chainId: (0, generic_1.networkByte)(paramsOrTx.chainId, 87),
        proofs: paramsOrTx.proofs || [],
        id: '',
    };
    validators_1.validate.commitToGeneraction(tx);
    const bytes = (0, proto_serialize_1.txToProtoBytes)(tx);
    seedsAndIndexes.forEach(([s, i]) => (0, generic_1.addProof)(tx, (0, ts_lib_crypto_1.signBytes)(s, bytes), i));
    tx.id = (0, ts_lib_crypto_1.base58Encode)((0, ts_lib_crypto_1.blake2b)(bytes));
    return tx;
}
