// @waves/waves-transactions/dist/transactions/issue.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.issue = issue;
/**
 * @module index
 */
const ts_lib_crypto_1 = require("@waves/ts-lib-crypto");
const marshall_1 = require("@waves/marshall");
const ts_types_1 = require("@waves/ts-types");
const generic_1 = require("../generic");
const validators_1 = require("../validators");
const proto_serialize_1 = require("../proto-serialize");
const defaultVersions_1 = require("../defaultVersions");
function issue(paramsOrTx, seed) {
    const type = ts_types_1.TRANSACTION_TYPE.ISSUE;
    const version = paramsOrTx.version || defaultVersions_1.DEFAULT_VERSIONS.ISSUE;
    const seedsAndIndexes = (0, generic_1.convertToPairs)(seed);
    const senderPublicKey = (0, generic_1.getSenderPublicKey)(seedsAndIndexes, paramsOrTx);
    const tx = {
        type,
        version,
        senderPublicKey,
        name: paramsOrTx.name,
        description: paramsOrTx.description,
        quantity: paramsOrTx.quantity,
        script: paramsOrTx.script == null ? null : (0, generic_1.base64Prefix)(paramsOrTx.script),
        decimals: paramsOrTx.decimals == null ? 8 : paramsOrTx.decimals,
        reissuable: paramsOrTx.reissuable || false,
        fee: checkForNFT(paramsOrTx) ? (0, generic_1.fee)(paramsOrTx, 100000) : (0, generic_1.fee)(paramsOrTx, 100000000),
        timestamp: paramsOrTx.timestamp || Date.now(),
        chainId: (0, generic_1.networkByte)(paramsOrTx.chainId, 87),
        proofs: paramsOrTx.proofs || [],
        id: '',
    };
    validators_1.validate.issue(tx);
    const bytes = version > 2 ? (0, proto_serialize_1.txToProtoBytes)(tx) : marshall_1.binary.serializeTx(tx);
    seedsAndIndexes.forEach(([s, i]) => (0, generic_1.addProof)(tx, (0, ts_lib_crypto_1.signBytes)(s, bytes), i));
    tx.id = (0, ts_lib_crypto_1.base58Encode)((0, ts_lib_crypto_1.blake2b)(bytes));
    return tx;
}
const checkForNFT = (paramsOrTx) => {
    return paramsOrTx.quantity === 1 && paramsOrTx.reissuable == false && paramsOrTx.decimals == 0;
};
