// @waves/waves-transactions/dist/transactions/lease.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.lease = lease;
/**
 * @module index
 */
const ts_lib_crypto_1 = require("@waves/ts-lib-crypto");
const marshall_1 = require("@waves/marshall");
const ts_types_1 = require("@waves/ts-types");
const generic_1 = require("../generic");
const validators_1 = require("../validators");
const proto_serialize_1 = require("../proto-serialize");
const defaultVersions_1 = require("../defaultVersions");
function lease(paramsOrTx, seed) {
    const type = ts_types_1.TRANSACTION_TYPE.LEASE;
    const version = paramsOrTx.version || defaultVersions_1.DEFAULT_VERSIONS.LEASE;
    const seedsAndIndexes = (0, generic_1.convertToPairs)(seed);
    const senderPublicKey = (0, generic_1.getSenderPublicKey)(seedsAndIndexes, paramsOrTx);
    const tx = {
        type,
        version,
        senderPublicKey,
        amount: paramsOrTx.amount,
        recipient: paramsOrTx.recipient,
        fee: (0, generic_1.fee)(paramsOrTx, 100000),
        timestamp: paramsOrTx.timestamp || Date.now(),
        proofs: paramsOrTx.proofs || [],
        chainId: (0, generic_1.networkByte)(paramsOrTx.chainId, 87),
        id: '',
    };
    validators_1.validate.lease(tx);
    const bytes = version > 2 ? (0, proto_serialize_1.txToProtoBytes)(tx) : marshall_1.binary.serializeTx(tx);
    seedsAndIndexes.forEach(([s, i]) => (0, generic_1.addProof)(tx, (0, ts_lib_crypto_1.signBytes)(s, bytes), i));
    tx.id = (0, ts_lib_crypto_1.base58Encode)((0, ts_lib_crypto_1.blake2b)(bytes));
    return tx;
}
