// @waves/waves-transactions/dist/requests/auth.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.serializeAuthData = void 0;
exports.auth = auth;
/**
 * @module index
 */
const ts_lib_crypto_1 = require("@waves/ts-lib-crypto");
const marshall_1 = require("@waves/marshall");
const generic_1 = require("../generic");
const validators_1 = require("../validators");
const { STRING, LEN, SHORT } = marshall_1.serializePrimitives;
const serializeAuthData = (auth) => (0, ts_lib_crypto_1.concat)(LEN(SHORT)(STRING)('WavesWalletAuthentication'), LEN(SHORT)(STRING)(auth.host || ''), LEN(SHORT)(STRING)(auth.data || ''));
exports.serializeAuthData = serializeAuthData;
function auth(params, seed, chainId) {
    const seedsAndIndexes = (0, generic_1.convertToPairs)(seed);
    const publicKey = params.publicKey || (0, generic_1.getSenderPublicKey)(seedsAndIndexes, { senderPublicKey: undefined });
    validators_1.validate.auth(params);
    const rx = {
        hash: '',
        signature: '',
        host: params.host,
        data: params.data,
        publicKey,
        address: (0, ts_lib_crypto_1.address)({ publicKey }, chainId),
    };
    const bytes = (0, exports.serializeAuthData)(rx);
    rx.signature = (seed != null && (0, ts_lib_crypto_1.signBytes)(seed, bytes)) || '';
    rx.hash = (0, ts_lib_crypto_1.base58Encode)((0, ts_lib_crypto_1.blake2b)(Uint8Array.from(bytes)));
    return rx;
}
